api/metrics/v1/http.go

```diff
@@ -526,6 +537,114 @@ func NewHandler(endpoints Endpoints, tlsOptions *tls.UpstreamOptions, opts ...Ha
 			))
 		})
 	}
+	if bkt != nil {
+		r.Group(func(r chi.Router) {
+			r.Use(server.StripTenantPrefix("/api/metrics/v1"))
+			r.Post(UploadRoute, func(w http.ResponseWriter, r *http.Request) {
```

Comment:
This route is doing a lot of processing, that could be done in the background.

Reply:
Yes, I agree we should do this async. We will need to decide if should we have external service like Kafka, rabbitmq where we can park these income request and process them asynchronously OR we have a simple go routine which will process the request in the background.

Reply:
Problem with doing that in a goroutine is that it's not durable. If the process dies for any reason (e.g. rolling out a new version) then the goroutine goes away and the blocks were never relabeled

Reply:
Yup, I meant doing this in some completely separate process instead is preferable.

Reply:
Can we do something like - 
When upload request is received we upload those `.tar.gz` archive to object store and than push a json message(containing a archive link) to a kafka topic. 
Than create a separate kafka consumer service which will consume these json message, download the `.tar.gz` file from object store and relabel the tsdb blocks and upload to the new object store bucket where Thanos store gateway is configured.